applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp

```diff
@@ -0,0 +1,743 @@
+                KRATOS_ERROR << "No submodelparts detected!" << std::endl;
+            
+            //initialize all required variables
+            rModelPart.GetProcessInfo()[BUILD_LEVEL] = 2;
+            SparseMatrixType* AuxStiffnessMatrix;
+            SparseMatrixPointerType temp_stiffness_matrix;
+            mMaterialDampingRatios = ZeroVector( n_modes );
+            
+            //initialize dummy vectors
+            auto pDx = SparseSpaceType::CreateEmptyVectorPointer();
+            auto pb = SparseSpaceType::CreateEmptyVectorPointer();
+            auto& rDx = *pDx;
+            auto& rb = *pb;
+            SparseSpaceType::Resize(rDx,system_size);
+            SparseSpaceType::Set(rDx,0.0);
+            SparseSpaceType::Resize(rb,system_size);
+            SparseSpaceType::Set(rb,0.0);
+
+            //loop over all modes and initialize the material damping ratio per mode
+            boost::timer material_damping_build_time;
+        
+            for( std::size_t i = 0; i < n_modes; ++i )
+            {
+                double up = 0.0;
+                double down = 0.0;
+                Vector modal_vector = column( rModalMatrix, i );
+                for( ModelPart::SubModelPartIterator itSubModelPart = rModelPart.SubModelPartsBegin(); itSubModelPart!= rModelPart.SubModelPartsEnd(); itSubModelPart++ )
+                {
+                    double damping_coefficient = 0.0;
+                    for( ModelPart::PropertiesIterator itProperty = itSubModelPart->PropertiesBegin(); itProperty != itSubModelPart->PropertiesEnd(); itProperty++ )
+                    {
+                        if( itProperty->Has(SYSTEM_DAMPING_RATIO) )
+                        {
+                            damping_coefficient = itProperty->GetValue(SYSTEM_DAMPING_RATIO);
+                        }
+                    }
+                    
+                    //initialize the submodelpart stiffness matrix
+                    AuxStiffnessMatrix = new SparseMatrixType;
+                    temp_stiffness_matrix = boost::shared_ptr<SparseMatrixType>(AuxStiffnessMatrix);
+                    pBuilderAndSolver->ResizeAndInitializeVectors(pScheme, 
+                        temp_stiffness_matrix,
+                        pDx,
+                        pb,
+                        rModelPart.Elements(),
+                        rModelPart.Conditions(),
+                        rModelPart.GetProcessInfo());
+
+                    //build stiffness matrix for submodelpart material
+                    pBuilderAndSolver->BuildLHS(pScheme,*itSubModelPart,*temp_stiffness_matrix);
+
+                    //compute strain energy of the submodelpart and the effective damping ratio
+                    double strain_energy = 0.5 * inner_prod( prod(modal_vector, *temp_stiffness_matrix), modal_vector );
+                    down += strain_energy;
+                    up += damping_coefficient * strain_energy;
+                }
+                if( down == 0 )
```

Comment:
do not compare a `double` with `==` => `if (down < std::numeric_limits epsilon)` (google for the exact syntax of numeric limits)